packages/contracts-bedrock/contracts/test/L1CrossDomainMessenger.t.sol

```diff
@@ -384,4 +387,307 @@ contract L1CrossDomainMessenger_Test is Messenger_Initializer {
         assertEq(L1Messenger.successfulMessages(hash), false);
         assertEq(L1Messenger.receivedMessages(hash), true);
     }
+
+    function test_relayMessage_legacy_succeeds() external {
+        address target = address(0xabcd);
+        address sender = Predeploys.L2_CROSS_DOMAIN_MESSENGER;
+
+        // Compute the message hash.
+        bytes32 hash = Hashing.hashCrossDomainMessageV1(
+            // Using a legacy nonce with version 0.
+            Encoding.encodeVersionedNonce(0, 0),
+            sender,
+            target,
+            0,
+            0,
+            hex"1111"
+        );
+
+        // Set the value of op.l2Sender() to be the L2 Cross Domain Messenger.
+        vm.store(address(op), bytes32(senderSlotIndex), bytes32(abi.encode(sender)));
+
+        // Target should be called with expected data.
+        vm.expectCall(target, hex"1111");
+
+        // Expect RelayedMessage event to be emitted.
+        vm.expectEmit(true, true, true, true);
+        emit RelayedMessage(hash);
+
+        // Relay the message.
+        vm.prank(address(op));
+        L1Messenger.relayMessage(
+            Encoding.encodeVersionedNonce(0, 0), // nonce
+            sender,
+            target,
+            0, // value
+            0,
+            hex"1111"
+        );
+
+        // Message was successfully relayed.
+        assertEq(L1Messenger.successfulMessages(hash), true);
+        assertEq(L1Messenger.receivedMessages(hash), false);
+    }
+
+    function test_relayMessage_legacyOldReplay_reverts() external {
+        address target = address(0xabcd);
+        address sender = Predeploys.L2_CROSS_DOMAIN_MESSENGER;
+
+        // Compute the message hash.
+        bytes32 hash = Hashing.hashCrossDomainMessageV1(
+            // Using a legacy nonce with version 0.
+            Encoding.encodeVersionedNonce(0, 0),
+            sender,
+            target,
+            0,
+            0,
+            hex"1111"
+        );
+
+        // Set the value of op.l2Sender() to be the L2 Cross Domain Messenger.
+        vm.store(address(op), bytes32(senderSlotIndex), bytes32(abi.encode(sender)));
+
+        // Mark legacy message as already relayed.
+        uint256 successfulMessagesSlot = 203;
```

Comment:
I've opened up an issue [here](https://github.com/foundry-rs/foundry/issues/4022) that would let us get rid of the magic numbers